### ISSUE TITLE
end to end training

### ISSUE THREAD

**Issue by VladKatsman** [JunwenBai/c-gmvae#6]
Guys, thank you for your paper and code.

The question is, if there any possibility for end to end efficient training. I mean train backbone (feature extractor) as well?

Would you suggest any efficient training scheme like next example:
training backbone first -> freeze backbone -> train g-cmvae -> unfreeze backbone -> finetune everything with smaller lr?

**Comment by roro5401** [comment]
# Questions about differences between the Code and Paper
Dear authors,

First of all, interesting paper with a great conceptual explanation of the framework. Thank you for making the code publicly available. I am looking to extend the model for my Master Thesis and I had a few questions about the code. Some things within the code are not very clear from the paper and vice versa. Answering these questiosn would greatly help my progress.

1. **Residual Connection:** I noticed in the code that the original input is concatenated to the latent represnetation before decoding into a feature embedding. I do not see any mention of this in the paper, is this a mistake in the code or intended but undiscussed? If the latter, have you ran any experiments without this residual connection and did it indeed improve performance?
2. **Data Views:** In the paper there is no mention of taking multiple samples from the latent space. I further see only one of the decoded views is used for computing the SupConLoss. What's the reason for not using all available views for the contrastive loss but using both views fo the BCE loss?
3. **SupConLoss:** Within the code for computing the SupConLoss you concatenate the reconstructed label emebddings to the feature embeddings while the paper mentions only the label embeddings are used as anchors. What is the point of this?
4. **Reconstruction of X:** The graphic in the paper shows x_hat as output of the moidel (presumably a reconstruction of the input). However, I do not see this in the code, or am I missing something?

Helping me clear this up would greatly improve my understanding of the code. 

Thanks in advance and I hope this reaches you well.

Robbert

**Comment by JunwenBai** [comment]
> Guys, thank you for your paper and code.
> 
> The question is, if there any possibility for end to end efficient training. I mean train backbone (feature extractor) as well?
> 
> Would you suggest any efficient training scheme like next example: training backbone first -> freeze backbone -> train g-cmvae -> unfreeze backbone -> finetune everything with smaller lr?

More elaborated finetuning process usually brings better performance. Given the trend of LLM and solid foundation models, it makes sense to have a solid backbone model. Regarding further unfreezing and re-finetuning, you may need to add some more experiments to validate that. One use case could be once you have more data or domains, you can unfreeze the backbone to further stretch. If the dataset remains intact, you may want to refine the backbone as much as possible within one stage.

**Comment by JunwenBai** [comment]
> # Questions about differences between the Code and Paper
> Dear authors,
> 
> First of all, interesting paper with a great conceptual explanation of the framework. Thank you for making the code publicly available. I am looking to extend the model for my Master Thesis and I had a few questions about the code. Some things within the code are not very clear from the paper and vice versa. Answering these questiosn would greatly help my progress.

Thx for the recognition and good luck with your thesis!

> 1. **Residual Connection:** I noticed in the code that the original input is concatenated to the latent represnetation before decoding into a feature embedding. I do not see any mention of this in the paper, is this a mistake in the code or intended but undiscussed? If the latter, have you ran any experiments without this residual connection and did it indeed improve performance?

The code has been iterated couple rounds for experimental variations and improvements. Residual connection is a common technique to bypass some information directly to ease the learning difficulty of the NN. I recall that this could be helpful.

> 2. **Data Views:** In the paper there is no mention of taking multiple samples from the latent space. I further see only one of the decoded views is used for computing the SupConLoss. What's the reason for not using all available views for the contrastive loss but using both views of the BCE loss?

Sry I don't quite understand your question. The feature embeddings are pulled together with the positive label embeddings, while pushing away from the negative label embeddings. The latent space is the posterior to be aligned with the prior. The single sample from Fig 1 is for illustration purpose.

> 3. **SupConLoss:** Within the code for computing the SupConLoss you concatenate the reconstructed label emebddings to the feature embeddings while the paper mentions only the label embeddings are used as anchors. What is the point of this?

We had experiments on both. If I recall correctly, both are good and this way may be slightly better for faster convergence and training stability.

> 4. **Reconstruction of X:** The graphic in the paper shows x_hat as output of the model (presumably a reconstruction of the input). However, I do not see this in the code, or am I missing something?> > Helping me clear this up would greatly improve my understanding of the code.> 

I think this question was raised earlier in other issues as well. Our initial impl had that and we surprisingly found that even skipping the reconstruction and keeping the contrastive loss alone could be effective too. So we checked in this simpler version. Feel free to add it back for you own experiments which should just add a few lines.
